src/execution/ingestion_sample.cpp

```diff
@@ -0,0 +1,680 @@
+#include "duckdb/common/types/data_chunk.hpp"
+#include "duckdb/common/vector_operations/vector_operations.hpp"
+#include "duckdb/execution/reservoir_sample.hpp"
+
+namespace duckdb {
+
+idx_t IngestionSample::NumSamplesCollected() const {
+	if (!sample_chunk) {
+		return 0;
+	}
+	return sample_chunk->size();
+}
+
+idx_t IngestionSample::NumActiveSamples() const {
+	auto ret = MaxValue<idx_t>(actual_sample_indexes.size(), base_reservoir_sample->reservoir_weights.size());
+	D_ASSERT(ret <= FIXED_SAMPLE_SIZE);
+	return ret;
+}
+
+unique_ptr<DataChunk> IngestionSample::GetChunk(idx_t offset) {
+	throw InternalException("Invalid Call to Get Chunk");
+}
+
+unique_ptr<DataChunk> IngestionSample::GetChunkAndShrink() {
+	throw InternalException("Invalid Sampling state");
+}
+
+unique_ptr<DataChunk> IngestionSample::CreateNewSampleChunk(vector<LogicalType> &types, idx_t size) const {
+	auto new_sample_chunk = make_uniq<DataChunk>();
+	new_sample_chunk->Initialize(Allocator::DefaultAllocator(), types, size);
+	for (idx_t col_idx = 0; col_idx < new_sample_chunk->ColumnCount(); col_idx++) {
+		auto type = types[col_idx];
+		// TODO: should the validity mask be the capacity or the size?
+		FlatVector::Validity(new_sample_chunk->data[col_idx]).Initialize(size);
+
+		if (!ValidSampleType(type)) {
+			new_sample_chunk->data[col_idx].SetVectorType(VectorType::CONSTANT_VECTOR);
+			ConstantVector::SetNull(new_sample_chunk->data[col_idx], true);
+		}
+	}
+
+	return new_sample_chunk;
+}
+
+SampleCopyHelper IngestionSample::GetSelToCopyData(idx_t sel_size) const {
+	SampleCopyHelper ret;
+	if (SamplingState() == SamplingMode::FAST) {
+		D_ASSERT(actual_sample_indexes.size() >= sel_size);
+		auto indexes_copy = actual_sample_indexes;
+		return SelFromSimpleIndexes(indexes_copy);
+	}
+	D_ASSERT(base_reservoir_sample->reservoir_weights.size() >= sel_size);
+	auto base_copy = base_reservoir_sample->Copy();
+	vector<std::pair<double, idx_t>> weights_indexes;
+	D_ASSERT(sel_size <= base_copy->reservoir_weights.size());
+	for (idx_t i = 0; i < sel_size; i++) {
+		weights_indexes.push_back(base_copy->reservoir_weights.top());
```

Comment:
Instead of keeping track of which rows are valid in the weights - perhaps we can add a `SelectionVector` that keeps track of which rows are valid in the sample. The weights can then keep on referring to the "logical row" (`0, 1, 2, 3, 4, 5, ...`) while the selection vector tracks the "actual row in the data chunk" (`17, 2, 3, ...`).